src/app/mod.rs

```diff
@@ -23,204 +26,412 @@ pub fn setup() {
+	///
+	/// Create a record with fields set
+	/// const record = {
+	///     name: 'Tobie',
+	///     settings: {
+	///         active: true,
+	///         marketing: true
+	///     }
+	/// };
+	/// const person = db.create('person', { content: record });
+	/// ```
+	pub async fn create(&self, resource: String, data: JsValue) -> Result<JsValue, Error> {
+		let resource = Resource::from(resource);
+		let response = match from_value::<Option<opt::resource::Create>>(data)? {
+			Some(data) => self.db.create(resource).content(data.content).await?,
+			None => self.db.create(resource).await?,
+		};
+		Ok(to_value(&Json::from(response))?)
 	}
 
-	#[wasm_bindgen]
-	pub fn update(&self, what: JsValue, data: JsValue) -> Promise {
-		match &self.kind {
-			Kind::None => Promise::reject(&JsValue::from(Error::MustInstantiate)),
-			Kind::Local(v) => future_to_promise(Local::update(v.clone(), what, data)),
-			Kind::Remote(_) => todo!(),
-		}
+	/// Update all records in a table, or a specific record
+	///
+	/// ```js
+	/// const record = {
+	///     name: 'Tobie',
+	///     settings: {
+	///         active: true,
+	///         marketing: true
+	///     }
+	/// };
+	///
+	/// // Replace the current document / record data with the specified data.
+	/// const person = db.update('person:tobie', { content: record });
+	///
+	/// // Merge the current document / record data with the specified data.
+	/// const person = db.update('person:tobie', { merge: record });
+	///
+	/// // Apply JSON Patch changes to all records, or a specific record, in the database.
+	/// const person = db.update('person:tobie', {
+	///     patch: { op: 'replace', path: '/settings/active', value: false }
+	/// });
+	/// ```
+	pub async fn update(&self, resource: String, data: JsValue) -> Result<JsValue, Error> {
+		let resource = Resource::from(resource);
+		let Some(data) = from_value::<Option<opt::resource::Update>>(data)? else {
+            let response = self.db.update(resource).await?;
+            return Ok(to_value(&Json::from(response))?);
+        };
+		let response = match (data.content, data.merge, data.patch) {
+			(None, None, None) => {
+				return Err("Expected content, merge or patch in the update method".into());
+			}
+			(Some(data), None, None) => self.db.update(resource).content(data).await?,
+			(None, Some(data), None) => self.db.update(resource).merge(data).await?,
+			(None, None, Some(mut patches)) => {
+				let mut patch = match patches.pop_front() {
+					Some(p) => self.db.update(resource).patch(match p {
+						Patch::Add {
+							path,
+							value,
+						} => PatchOp::add(&path, value),
+						Patch::Remove {
+							path,
+						} => PatchOp::remove(&path),
+						Patch::Replace {
+							path,
+							value,
+						} => PatchOp::replace(&path, value),
+						Patch::Change {
+							path,
+							diff,
+						} => PatchOp::change(
+							&path,
+							Diff {
+								operation: diff.operation,
+								text: diff.text,
+							},
+						),
+					}),
```

Comment:
Can you explain what is happening here? Why are we taking the first patch entry, and why are we looping over them in the code below?

Also on a side note, can we pass the handling of this through to the Rust API library itself, as opposed to handling this here in the WASM library?

Reply:
In the Rust API we can chain patches like so

```rust
db.update(..)
    .patch(..)
    .patch(..)
    .await?;
```

We can't do this in JS so we have to pass in all the patches at once

```js
await db.update('..', [
    { op: 'add', .. },
    { op: 'replace', .. }
]);
```

That plumbing is just there to translate the JS vector approach to Rust's chained method approach and also translate the intermediary type deserialised into from JS to Rust's `PatchOp`.

We extract the first element before looping over the rest so we can use the first patch to setup the correct type that Rust would expect.

Reply:
I added some comments to the code.